File: alias.go
package main

import (
	"fmt"
	"strings"
)

func parseAliases(rows []string) (map[string]string, map[string]string) {
	if len(rows) == 0 {
		return nil, nil
	}

	res := make(map[string]string, len(rows))
	viceVersa := make(map[string]string, len(rows))

	for _, row := range rows {
		parts := strings.Split(row, "=")

		if len(parts) < 2 {
			continue
		}

		alias, command := extract(parts[0]), extract(parts[1])

		res[command] = extract(alias)
		viceVersa[alias] = extract(command)
	}

	return res, viceVersa
}

func find(cmd string, list map[string]string) (value string, found bool) {
	cmd = strings.TrimSpace(cmd)

	if len(cmd) == 0 || len(list) == 0 {
		return
	}

	parts := strings.Split(cmd, "&&")
	if len(parts) > 1 {
		res := make([]string, len(parts))
		for i := range parts {
			parts[i] = strings.TrimSpace(parts[i])

			a, f := find(parts[i], list)
			if f {
				found = true
				res[i] = a
			} else {
				res[i] = parts[i]
			}
		}
		return strings.Join(res, " && "), found
	}

	value, found = list[cmd]
	if !found {
		var args string

		args, rest := pop(strings.Split(cmd, " "))

		value, found = find(strings.Join(rest, " "), list)
		if !found {
			return
		}

		value = fmt.Sprintf("%s %s", value, args)

		return
	}

	return
}

func extract(str string) string {
	if len(str) == 0 || str[0] != '\'' {
		return str
	}

	return str[1 : len(str)-1]
}

func pop(arr []string) (string, []string) {
	if len(arr) == 0 {
		return "", nil
	}

	lastIx := len(arr)

	return arr[lastIx-1], arr[:lastIx-1]
}

File: alias.py
def parse_aliases(rows):
    by_cmd, by_alias = {}, {}

    for row in rows:
        parts = row.split("=")
        if len(parts) != 2:
            continue

        alias, command = extract(parts[0]), extract(parts[1])

        by_cmd[command] = alias
        by_alias[alias] = command

    return by_cmd, by_alias

def find_alias(cmd, items):
    if len(cmd) == 0 or len(items) == 0:
        return None

    if cmd in items:
        return items[cmd]

    sorted_items = dict(sorted(items.items(), reverse=True))

    for full, alias in sorted_items.items():
        if cmd.find(full) == 0:
            return cmd.replace(full, alias)

    return None

def suggest_alias(cmd, freq, history):
    freq = max(freq, 1)
    freq = min(freq, 100)

    if len(cmd) == 0 or len(history) == 0:
        return None

    count = 1
    for row in history:
        row = " ".join(" ".split(row.strip())[1:]).strip()
        if row == cmd:
            count += 1

    call_percent = count * 100 / (len(history) + 1)
    if call_percent >= freq:
        return cmd

    return None

def extract(str):
    if len(str) == 0 or str[0] != "'":
        return str

    return str[1:len(str) - 1]
